src/driver/buffers.rs

```diff
@@ -0,0 +1,132 @@
+use crate::runtime::CONTEXT;
+
+use libc::{iovec, UIO_MAXIOV};
+use std::cell::RefCell;
+use std::cmp;
+use std::io;
+use std::mem;
+use std::ptr;
+use std::rc::Rc;
+use std::slice;
+
+pub(crate) struct Buffers {
+    raw_vecs: ptr::NonNull<iovec>,
+    buf_states: Vec<BufState>,
+    orig_cap: usize,
+}
+
+enum BufState {
+    Free { init_len: usize },
+    CheckedOut,
```

Comment:
When the buffer is checked out, the length of the initialized part is recorded in the `FixedBuf`. It can also be changed by operations, so it's best to have it accounted in a single place.

Reply:
Makes sense. Would you mind adding a one line comment to that effect for this invariant?